### PR TITLE
Issue 116

### PR DESCRIPTION
**Pull request by aleh-null**

### CODE REVIEW

File: pkg/devcontainer/image_manager.go [hunk 1]
```diff
@@ -31,6 +32,15 @@ func NewImageManager(dockerImageCli client.ImageAPIClient, randomString func(int
...
 
+	imgs, err := im.ImageList(ctx, image.ListOptions{Filters: filters.NewArgs(filters.Arg("reference", name))})
+	if err != nil {
+		return fmt.Errorf("Failed to list local images, %w", err)
+	}
+	if len(imgs) != 0 {
+		log.Debug().Str("image", name).Msg("Local image exists")
+		return nil
+	}
+
```
Comment (artmoskvin) [inline pkg/devcontainer/image_manager.go:43]: i think this logic deserves a separate method that we can call from the devcontainer manager and pull the image only if local not found.

File: pkg/jsonc/jsonc_test.go [hunk 2]
```diff
@@ -26,3 +26,13 @@ func TestToJSON(t *testing.T) {
 		t.Fatalf("expected '%s', got '%s'", expect, out)
 	}
 }
+
+func TestToJSON2(t *testing.T) {
```
Comment (artmoskvin) [inline pkg/jsonc/jsonc_test.go:30]: @aleh-null do we still need this test?

File: pkg/devcontainer/image_manager.go [hunk 3]
```diff
@@ -138,6 +130,20 @@ func (im *DockerImageManager) BuildImage(ctx context.Context, workingDir string,
 	return tag, nil
 }
 
+// CheckLocalImage checks if an image with the given name exists locally.
+// It returns true if the image exists, false if it doesn't, and an error if the check fails.
+func (im *DockerImageManager) CheckLocalImage(ctx context.Context, name string) (bool, error) {
```
Comment (aleh-null) [inline pkg/devcontainer/image_manager.go:135]: rename `HasLocalImage()` -> `LocalImageExists()`

**Review by aleh-null — COMMENTED** [review]
Just a rename suggesting. Otherwise LGTM.

File: pkg/devcontainer/runner.go [hunk 4]
```diff
@@ -157,3 +135,34 @@ func (r *DockerRunner) executeLifecycleCommandInContainer(ctx context.Context, l
 
 	return nil
 }
+
+func (r *DockerRunner) pullOrBuildImage(ctx context.Context, config Config, projectPath string) (string, error) {
```
Comment (aleh-null) [inline pkg/devcontainer/runner.go:139]: Hm, we do more that that. Maybe we should just abstract `case config.IsImageDevContainer()`

File: pkg/jsonc/jsonc_test.go (hunk 2 above)
Comment (aleh-null) [reply pkg/jsonc/jsonc_test.go:30]: Probably a good idea to have test on the actual dev container. WDYT of using https://github.com/marcozac/go-jsonc in the future?

Comment (artmoskvin) [reply pkg/jsonc/jsonc_test.go:30]: is it better?

File: pkg/devcontainer/runner.go (hunk 4 above)
Comment (artmoskvin) [reply pkg/devcontainer/runner.go]: extracted the `IsImageDevContainer` case in a separate method. we can abstract the whole image thing further into something like `ImageProvider` but i think it's an overkill for now.

File: pkg/devcontainer/image_manager.go (hunk 3 above)
Comment (artmoskvin) [reply pkg/devcontainer/image_manager.go]: done

**Review by artmoskvin — APPROVED** [review]

**Pull request merged**